### PR TITLE
Fixed selection changes with arrow keys

### LINKED ISSUE #9380: An inline image in the table blocks the cross-cell navigation.
## 📝 Provide detailed reproduction steps (if any)

1. Open the `i/2052-inline-images` branch.
2. Run and open the all-features manual test.
3. Set the cursor at the end of the text in a table cell and insert an image.
4. Select the image and try to navigate to the next cell on the right/bottom.
5. Set the cursor at the beginning of the text in a table cell and insert an image.
6. Select the image and try to navigate to the previous cell on the left/top.

### ✔️ Expected result

It should be possible to go to the previous/next cell with the keyboard.

### ❌ Actual result

The selection returns to the inline image instead of moving to the next/previous cell.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### LINKED ISSUE #10391: Non-collapsed selection should not jump to another table cell on arrow key press
## 📝 Provide detailed reproduction steps (if any)

1.  Select content of some table cell
2.  press any arrow key

### ✔️ Expected result

Selection should get collapsed

### ❌ Actual result

Selection moves to another cell

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by niegowski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (widget): Fixed arrow keys navigation when there is an inline widget at the edge of a table cell. Closes #9380.

Fix (table, widget): Fixed handling of non-collapsed selection inside a table cell. Closes #10391.

---

### Additional information

### CODE REVIEW

**Comment by Reinmar** [comment]
cc @ckeditor/qa-team Could you take a look? This needs to be verified for cross-browser compatibility.

**Comment by Mgsy** [comment]
### Firefox

*   Go to http://localhost:8125/ckeditor5/tests/manual/all-features.html.
*   Set the following data:

```
<p>&nbsp;</p><figure class=\"table\" style=\"width:600px;\"><table style=\"border:2px dashed hsl(0, 0%, 60%);\"><tbody><tr><td style=\"vertical-align:top;width:33%;\"><p><img class=\"image_resized\" style=\"width:48.31%;\" src=\"sample.jpg\" alt=\"bar\"></p><p>L List item 1</p><ul><li>UL List item 2</li></ul><figure class=\"image image_resized\" style=\"width:48.31%;\"><img src=\"sample.jpg\" alt=\"bar\"></figure></td><td style=\"vertical-align:top;width:33%;\">&nbsp;</td><td style=\"vertical-align:top;width:33%;\">&nbsp;</td></tr></tbody></table></figure>
```

*   Put the caret in the first cell and select all.
*   Press Arrow right/left.

### Actual result

The selection doesn't collapse.

It doesn't occur without changes introduced by this PR.

**Comment by niegowski** [comment]
The above case should be fixed now, @ckeditor/qa-team could you take another look?

**Comment by Mgsy** [comment]
It still could be reproduced:

*   Go to the all-features manual test.
*   Call:

```
editor.setData( "<figure class=\"table\" style=\"width:600px;\"><table style=\"border:2px dashed hsl(0, 0%, 60%);\"><tbody><tr><td style=\"vertical-align:top;width:33%;\"><p><img class=\"image_resized\" style=\"width:63.83%;\" src=\"sample.jpg\" alt=\"bar\"></p><ul><li>UL List item 1</li><li>UL List item 2</li></ul><p><img class=\"image_resized\" style=\"width:36.55%;\" src=\"sample.jpg\" alt=\"bar\"></p></td><td style=\"vertical-align:top;width:33%;\">&nbsp;</td><td style=\"vertical-align:top;width:33%;\">&nbsp;</td></tr></tbody></table></figure>" )
```

*   Focus the first cell.
*   Select all and press Arrow right/left.

**Comment by niegowski** [comment]
Looks like browsers have issues with collapsing selections that start or end with any widget (no matter if inline or block and on which end of the selection). A similar issue happened in Chrome - selection was jumping in strange positions.

@ckeditor/qa-team please check it once again. I hope this time it's finally fixed 🤞

File: packages/ckeditor5-table/src/tablekeyboard.js [hunk 1]
```diff
@@ -218,10 +218,24 @@ export default class TableKeyboard extends Plugin {
...
-		// Selection for sure will not approach cell edge.
-		if ( expandSelection && !selection.isCollapsed && selection.isBackward == isForward ) {
-			return false;
+		// When the selection is not collapsed.
+		if ( !selection.isCollapsed ) {
+			if ( expandSelection ) {
+				// Navigation is in the opposite direction than the selection direction so this is shrinking of the selection.
+				// Selection for sure will not approach cell edge.
+				//
+				// With a special case when all cell content is selected - then selection should expand to the other cell.
```
Comment (oleq) [inline packages/ckeditor5-table/src/tablekeyboard.js:227]:
```suggestion
				// With a special case when all cell content is selected - then selection should expand to the other cell.
				// Note: When the entire cell gets selected using CTRL+A, the selection is always forward.				
```

**Review by oleq — APPROVED** [review]

**Pull request merged**